lib/Doctrine/DBAL/TransactionManager.php

```diff
@@ -0,0 +1,284 @@
+<?php
+
+namespace Doctrine\DBAL;
+
+/**
+ * Manages transactions.
+ */
+class TransactionManager
+{
+    const SAVEPOINT_PREFIX = 'DOCTRINE2_SAVEPOINT_';
+
+    /**
+     * @var \Doctrine\DBAL\Connection
+     */
+    private $connection;
+
+    /**
+     * @var \Doctrine\DBAL\Transaction[]
+     */
+    private $activeTransactions = array();
+
+    /**
+     * Whether nested transactions should use savepoints.
+     *
+     * @var boolean
+     */
+    private $nestTransactionsWithSavepoints = false;
+
+    /**
+     * @param \Doctrine\DBAL\Connection $connection
+     */
+    public function __construct(Connection $connection)
+    {
+        $this->connection = $connection;
+    }
+
+    /**
+     * @return TransactionBuilder
+     */
+    public function createTransaction()
+    {
+        return new TransactionBuilder($this);
+    }
+
+    /**
+     * @param array $configuration
+     *
+     * @return \Doctrine\DBAL\Transaction
+     */
+    public function beginTransaction(array $configuration = array())
+    {
+        $this->connection->connect();
+
+        if (isset($configuration[TransactionBuilder::ISOLATION_LEVEL])) {
+            $this->connection->setTransactionIsolation($configuration[TransactionBuilder::ISOLATION_LEVEL]);
+        }
+
+        $logger = $this->connection->getConfiguration()->getSQLLogger();
+
+        if (count($this->activeTransactions) === 0) {
+            $logger && $logger->startQuery('"START TRANSACTION"');
+            $this->connection->getWrappedConnection()->beginTransaction();
```

Comment:
As far as I understand, the wrapper connection is called to bypass the deprecated method which would call the transaction manager back. Is it a temporary solution or permanent?

Reply:
True. We're going to remove the current implementation of `Connection::beginTransaction()` anyway, so at some point we need to actually begin the transaction on the underlying *driver* connection.

This is supposed to be permanent. This may make more sense if we decide to merge Connection and TransactionManager. BTW, I think `getWrappedConnection()` should be renamed to `getDriverConnection()`.

Reply:
As a permanent solution, it still violates the Law of Demeter. As long as the transaction manager interacts with a connection object, it should interact with it directly without depending on the interface of the internal connection.

Reply:
So I believe this is a point in favour of merging Connection and TransactionManager then?

Reply:
It would solve the problem, yeh.